src/plans/nonlinear_least_squares_plan.jl

```diff
@@ -0,0 +1,209 @@
+
+@doc raw"""
+    NonlinearLeastSquaresProblem{T<:AbstractEvaluationType} <: Problem{T}
+
+A type for nonlinear least squares problems.
+`T` is a [`AbstractEvaluationType`](@ref) for the `F` and Jacobian functions.
+
+Specify a nonlinear least squares problem
+
+# Fields
+* `M`        – a manifold ``\mathcal M``
+* `F`        – a function ``F: \mathcal M → ℝ^d`` to minimize
+* `jacF!!`   – Jacobian of the function ``F``
+* `jacB`     – the basis of tangent space used for computing the Jacobian.
+* `num_components` – number of values returned by `F` (equal to `d`).
+
+Depending on the [`AbstractEvaluationType`](@ref) `T` the function ``F`` has to be provided:
+
+* as a functions `(M::AbstractManifold, p) -> v` that allocates memory for `v` itself for
+  an [`AllocatingEvaluation`](@ref),
+* as a function `(M::AbstractManifold, v, p) -> v` that works in place of `v` for a
+  [`MutatingEvaluation`](@ref).
+
+Also the Jacobian ``jacF!!`` is required:
+
+* as a functions `(M::AbstractManifold, p; B_dom::AbstractBasis) -> v` that allocates memory
+  for `v` itself for an [`AllocatingEvaluation`](@ref),
+* as a function `(M::AbstractManifold, v, p; B_dom::AbstractBasis) -> v` that works in place
+  of `v` for an [`MutatingEvaluation`](@ref).
+
+# Constructors
+
+    NonlinearLeastSquaresProblem(M, F, jacF, num_components; evaluation=AllocatingEvaluation(), jacB=DefaultOrthonormalBasis())
+
+# See also
+
+[`LevenbergMarquardt`](@ref), [`LevenbergMarquardtOptions`](@ref)
+"""
+struct NonlinearLeastSquaresProblem{
+    T<:AbstractEvaluationType,mT<:AbstractManifold,TF,TJ,TB<:AbstractBasis
+} <: Problem{T}
+    M::mT
+    F::TF
+    jacobian!!::TJ
+    jacB::TB
+    num_components::Int
+end
+function NonlinearLeastSquaresProblem(
+    M::mT,
+    F::TF,
+    jacF::TJ,
+    num_components::Int;
+    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
+    jacB::TB=DefaultOrthonormalBasis(),
+) where {mT<:AbstractManifold,TF,TJ,TB<:AbstractBasis}
+    return NonlinearLeastSquaresProblem{typeof(evaluation),mT,TF,TJ,TB}(
+        M, F, jacF, jacB, num_components
+    )
+end
+
+function (d::DebugGradient)(::NonlinearLeastSquaresProblem, o::Options, i::Int)
+    (i < 1) && return nothing
+    Printf.format(d.io, Printf.Format(d.format), get_gradient(o))
+    return nothing
+end
+
+function get_cost(P::NonlinearLeastSquaresProblem{AllocatingEvaluation}, p)
+    return 1//2 * norm(P.F(P.M, p))^2
+end
+function get_cost(P::NonlinearLeastSquaresProblem{MutatingEvaluation}, p)
+    Fval = zeros(P.num_components)
+    P.F(P.M, Fval, p)
+    return 1//2 * norm(Fval)^2
+end
```

Comment:
What exactly would `get_cost!` do?

Reply:
`get_cost!(P, Fval, p)` would evaluate the (vectorial) cost in place, similarly we could use a `get_cost(P, p)` to allocate (which is done here anyways).

Reply:
I'm a little worried that it might be confusing as that's not really cost but the residual, with the cost being half of the norm of the residual.

Also note that to simplify
```julia
    if Teval === AllocatingEvaluation
        o.candidate_cost_values = p.F(p.M, temp_x)
    else
        p.F(p.M, o.candidate_cost_values, temp_x)
    end
```
we would have to pass entire options structure to such `get_cost`, which is not a pattern I've seen in Manopt so far.

Reply:
Then let's call it `get_residuals`(I thought it was costs) - also surprising that it is then stored in something that is called costs if it is not costs (yes I proposed the name but for the same reason)

Concerning the passing – no? 

a) would be `o.candidate_cost_values = get_residuals(p, temp_x)`
b) would be `get_residuals!(p, o.candidate_cost_values, temp_x)`

and actually I would just use a) and have 4 variants of `get_residuals` / `get_residuals!` (two each, namely one for Allocating one for Mutating for each of the functions).

Reply:
> Then let's call it `get_residuals`(I thought it was costs) - also surprising that it is then stored in something that is called costs if it is not costs (yes I proposed the name but for the same reason)

Maybe calling it costs wasn't a particularly good choice. Nonlinear least squares is primarily used for regression-type problems where `F` returns residuals. In particular it has nothing to do with multi-objective optimization.

> Concerning the passing – no?
> 
> a) would be `o.candidate_cost_values = get_residuals(p, temp_x)` b) would be `get_residuals!(p, o.candidate_cost_values, temp_x)`
> 
> and actually I would just use a) and have 4 variants of `get_residuals` / `get_residuals!` (two each, namely one for Allocating one for Mutating for each of the functions).

Yes, that would work but would
```julia
    if Teval === AllocatingEvaluation
        o.candidate_cost_values = get_residuals(p, temp_x)
    else
        get_residuals!(p, o.candidate_cost_values, temp_x)
    end
```
be much nicer than the current code?

Reply:
That won't work if `o.candidate_cost_values` is for example `SArray`, which I think would be nice to support.

Reply:
hm, but then nearly all of my accessor function have this problem at the moment, we should talk about that – maybe even in a small meeting on zoom?

Reply:
Because until now I thoght the approach with 4 functions – sure it might allocate for the allocating variant, but there is nothing one can do about that; I am referring to the 4 functions like

https://github.com/JuliaManifolds/Manopt.jl/blob/40ec55022f3aefc51ae28347c123621cbaa4cd02/src/plans/gradient_plan.jl#L54-L68

which encapsulate nicely whether its mutating or not – and the user can use this or that variant (`get_gradient`/ `get_gradient!`) to indicate basically “if this is possible, just mutate do not allocate” and the `!` variant for allocating functions – well still allocates and then copies.

Reply:
> hm, but then nearly all of my accessor function have this problem at the moment, we should talk about that – maybe even in a small meeting on zoom?

It's not a big issue but it would be nice to use `SArray` for gradients, for example, in the allocating evaluation mode. We can schedule a zoom meeting on Slack if you want :slightly_smiling_face: .

Reply:
In the latest commit I've sketched a possible alternative we've discussed. What do you think? I know `access_field(o, :jacF, Teval())` isn't particularly readable but there is not that much that can be done. Alternatively we could have `(@access_field o.jacF)` with `Teval` being implicitly captured.

Reply:
For me that does look ok, I just have not yet understood where we need the `FieldAccessor` or another words how complicated it is for the user to setup the `o.jacF` now ...

Ah, I think I got it, so basically for the allocating case, we create an `FieldAcessor` which does the job (maybe comparable to a view?) so

a) for the user it would enable to _just_ use it as before?
b) we internally need this small trick with `acess_field`? And alternatively we could pass `o` to the `get_jacobian(!)` and do the access there?

Reply:
I think I would need these functions and types properly documented to understand the last few details but besides that the `access_field` seems a good solution and the macro looks even nicer if that would be possible, then we are nearly back to classical code from before.

Reply:
Yes, the user sees no difference and people writing new methods just need to remember to use `access_field`. And `FieldAccessor` is actually more comparable to a reference (`Ref`). I will add some documentation and the macro.

Reply:
Ah yes, but also `Ref` us basically a “short-time-wrap”. Thanks for that idea, looking forward to using that. I will wait on the other branch for this to be finished, to directly merge and adopt it then :)

Reply:
Cool :slightly_smiling_face: . I've added some docs, macro and changed the type name to `FieldReference` to make it seem more related to `Ref`.

Reply:
This looks quite cool – is nearly as easy as the old variant and with the macro still very nice within the algorithms! Thanks.